packages/admin-ui/src/number-input/number-input.state.ts

```diff
@@ -0,0 +1,115 @@
+import { useEffect, useState } from 'react'
+import type { ButtonProps } from '../button'
+import {
+  isNumber,
+  isNumberAsString,
+  getIncrementedValue,
+  getDecrementedValue,
+} from './utils'
+
+export const useNumberInput = (props: NumberInputState) => {
+  const { initialValue, step, min, max, onChange } = props
+  const [currentValue, setCurrentValue] = useState(initialValue)
+
+  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (event) => {
+    const { value } = event.target
+    const newValue = isNumberAsString(value) ? Number(value) : value
+
+    setCurrentValue(newValue)
+    onChange(newValue)
+  }
+
+  const handleIncrement: React.MouseEventHandler<HTMLButtonElement> = () => {
+    if (isNumber(currentValue)) {
+      const newValue = getIncrementedValue({
+        currentValue,
+        nextValue: currentValue + step,
+        min,
+        max,
+      })
+
+      setCurrentValue(newValue)
+      onChange(newValue)
+    } else {
+      // When the currentValue is a empty string and you click to increment
+      const newValue = max <= step ? max : step
+
+      setCurrentValue(newValue)
+      onChange(newValue)
+    }
+  }
+
+  const handleDecrement: React.MouseEventHandler<HTMLButtonElement> = () => {
+    if (isNumber(currentValue)) {
+      const newValue = getDecrementedValue({
+        currentValue,
+        nextValue: currentValue - step,
+        min,
+        max,
+      })
+
+      setCurrentValue(newValue)
+      onChange(newValue)
+    } else {
+      // When the currentValue is a empty string and you click to decrement
```

Comment:
I think it would be better for the default value to be always 0, this way you don't need to make this condition of an empty string. What do you think? Maybe is worth talking to davis